recipes/opentelemetry-cpp/all/conanfile.py

```diff
@@ -52,6 +54,8 @@ class OpenTelemetryCppConan(ConanFile):
         "with_otlp": "deprecated",
         "with_otlp_grpc": True,
         "with_otlp_http": True,
+        "with_otlp_http_compression": False,
+        "with_otlp_file": False,
```

Comment:
Based on https://github.com/open-telemetry/opentelemetry-cpp/compare/v1.14.2...v1.15.0#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR376 and follwoing lines, I think these options should influence which requirements are pulled!

Reply:
@RubenRBS 
Thank you for your comment!
I fixed this PR for protobuf and zlib.

I'm also trying to incorporate #22617 improvements.

Reply:
```
if(WITH_OTLP_GRPC
   OR WITH_OTLP_HTTP
   OR WITH_OTLP_FILE)
  find_package(Protobuf)
```
https://github.com/open-telemetry/opentelemetry-cpp/blob/v1.15.0/CMakeLists.txt#L376-L379


```
if(WITH_OTLP_HTTP
   OR WITH_ELASTICSEARCH
   OR WITH_ZIPKIN
   OR BUILD_W3CTRACECONTEXT_TEST
   OR WITH_EXAMPLES_HTTP)
  set(WITH_HTTP_CLIENT_CURL ON)
else()
  set(WITH_HTTP_CLIENT_CURL OFF)
endif()
```
https://github.com/open-telemetry/opentelemetry-cpp/blob/v1.15.0/CMakeLists.txt#L442-L450


```
if((NOT WITH_API_ONLY)
   AND WITH_HTTP_CLIENT_CURL
   AND WITH_OTLP_HTTP_COMPRESSION)
  # No specific version required.
  find_package(ZLIB REQUIRED)
  message(STATUS "Found ZLIB: ${ZLIB_LIBRARIES}, version ${ZLIB_VERSION}")
endif()
```
https://github.com/open-telemetry/opentelemetry-cpp/blob/v1.15.0/CMakeLists.txt#L466-L472

Reply:
Just chiming in to say that I would particularly appreciate porting the fixes to the `WITH_STL` specification from #22617

Reply:
@phantom-jacob 
I try to merge #22617 improvements as possible.
Could you review this please?

Reply:
I'm not sure what's causing the check failures, but the changes around `WITH_STL` look like they'll solve my local configuration issue. My problem with the current production recipe is that I can't get `opentelemetry-cpp` to use the stdlib, because the `WITH_STL` configuration currently available to conanfiles is insufficient to get the package to use stdlib. The changes which set the `WITH_STL` config to a specific value, and the `OPENTELEMETRY_STL_VERSION` macro to the corresponding correct value, _should_ make it possible to do what I want.

I'm not sure I know the right procedure to test a conan recipe off an open PR though

Reply:
@toge I'm seeing errors in builds that seem related to C++ std selection. It would be a smaller change, and still would solve the underlying `WITH_STL` issue, if you were to simply update the `with_stl` option to support `CXX` strings, leaving the user to manually select their C++ standard instead of trying to infer it. This might make it easier to pass tests, while still fixing the usability problem

Reply:
For example, I believe this diff would be enough to meet my needs:
```diff
diff --git a/recipes/opentelemetry-cpp/all/conanfile.py b/recipes/opentelemetry-cpp/all/conanfile.py
index bb02fc972..ae64f5fb9 100644
--- a/recipes/opentelemetry-cpp/all/conanfile.py
+++ b/recipes/opentelemetry-cpp/all/conanfile.py
@@ -24,7 +24,7 @@ class OpenTelemetryCppConan(ConanFile):
         "fPIC": [True, False],
         "shared": [True, False],
         "with_no_deprecated_code": [True, False],
-        "with_stl": [True, False],
+        "with_stl": [True, False, "CXX14", "CXX17", "CXX20", "CXX23"],
         "with_gsl": [True, False],
         "with_abseil": [True, False],
         "with_otlp": ["deprecated", True, False],
@@ -90,28 +90,6 @@ class OpenTelemetryCppConan(ConanFile):
                 "msvc": "192",
             }

-    @property
-    def _used_cppstd(self):
-        return self.settings.compiler.get_safe("cppstd") or self._min_cppstd
-
-    @property
-    def _with_stl_value(self):
-        if Version(self.version) >= "1.12":
-            if self.options.with_stl:
-                cppstd = self._used_cppstd
-                if "14" in cppstd:
-                    return "CXX14"
-                elif "17" in cppstd:
-                    return "CXX17"
-                elif "20" in cppstd:
-                    return "CXX20"
-                elif "23" in cppstd:
-                    return "CXX23"
-            else:
-                return "OFF"
-        else:
-            return self.options.with_stl
-
     def config_options(self):
         if self.settings.os == "Windows":
             self.options.rm_safe("fPIC")
@@ -248,7 +226,7 @@ class OpenTelemetryCppConan(ConanFile):
         tc.cache_variables["BUILD_BENCHMARK"] = False
         tc.cache_variables["WITH_EXAMPLES"] = False
         tc.cache_variables["WITH_NO_DEPRECATED_CODE"] = self.options.with_no_deprecated_code
-        tc.cache_variables["WITH_STL"] = self._with_stl_value
+        tc.cache_variables["WITH_STL"] = self.options.with_stl
         tc.cache_variables["WITH_GSL"] = self.options.with_gsl
         tc.cache_variables["WITH_ABSEIL"] = self.options.with_abseil
         if Version(self.version) < "1.10":
@@ -413,9 +391,9 @@ class OpenTelemetryCppConan(ConanFile):
         if self.settings.os in ("Linux", "FreeBSD"):
             self.cpp_info.components["opentelemetry_common"].system_libs.extend(["pthread"])

-        if self.options.with_stl:
+        if isinstance(self.options.with_stl, str):
             if Version(self.version) >= "1.12.0":
-                cppstd = self._used_cppstd[-2:]
+                cppstd = self.options.with_stl[-2:]
                 self.cpp_info.components["opentelemetry_common"].defines.append(f"OPENTELEMETRY_STL_VERSION=20{cppstd}")
             else:
                 self.cpp_info.components["opentelemetry_common"].defines.append("HAVE_CPP_STDLIB")
```

That said, I'm definitely a novice w.r.t. conan, so feel free to disregard if I'm off base 😄